### PR TITLE
Make TimeSeriesLifecycleActionsIT.testWaitForSnapshot and testWaitForSnapshotSlmExecutedBefore wait for snaphost

### PR DESCRIPTION
**Pull request by probakowski**
This change modifies `TimeSeriesLifecycleActionsIT.testWaitForSnapshot` and `testWaitForSnapshotSlmExecutedBefore` to wait for snapshots to execute successfully. 
This gives more time for test to finish (as there are more `assertBusy` calls) and makes debugging easier as allows to differentiate between problem with snapshot and problem with `WaitForSnapshotStep`.
Additionally new assertions should fix race condition between SLM policy execution and ILM recovering retriable step from error.
I've also added debug informations (explain index and SLM policy data) to ease debugging and changed snapshot repo location used by these 2 tests to random to avoid clashes with other snapshots.
I would like to merge it to `master` only for now and if it stays quite for some time I'd backport it to `7.x`

Reletes to #50781

### CODE REVIEW

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by probakowski** [comment]
@elasticmachine update branch

**Review by dakrone — APPROVED** [review]
LGTM

**Comment by probakowski** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 1]
```diff
@@ -328,64 +329,67 @@ public void testAllocateActionOnlyReplicas() throws Exception {
...
+                Map<String, Object> responseMap = XContentHelper.convertToMap(XContentType.JSON.xContent(), is, true);
+                assertEquals(1, ((Map<?, ?>) ((Map<?, ?>) responseMap.get(slmPolicy)).get("stats")).get("snapshots_taken"));
+            }
+        }, slmPolicy);
 
-        request = new Request("PUT", "/_slm/policy/" + smlPolicy + "/_execute");
-        assertOK(client().performRequest(request));
+        updatePolicy(index, policy);
+        assertBusy(() -> assertThat(getStepKeyForIndex(index).getAction(), equalTo("wait_for_snapshot")), slmPolicy);
+        assertBusy(() -> assertThat(getStepKeyForIndex(index).getName(), equalTo("wait-for-snapshot")), slmPolicy);
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: I believe these 2 can be grouped together in one `assertBusy` call as they both make assertions about the same index state (that could also reduce the GET calls we make as `getStepKeyForIndex` could only be called once)

I think even the `getFailedStepForIndex` below could be grouped into the same `assertBusy` call. 

I think something along the lines of 
```
assertBusy( () -> {
    Map<String, Object> indexILMState = explainIndex(index);
    assertThat(indexILMState.get("action"), is("wait_for_snapshot");
    assertThat(indexILMState.get("step"), is("wait-for-snapshot");
    assertThat(indexILMState.get("failed_step"), nullValue());
});
```
will reduce the number of HTTP calls we make and the time we spend on these assertions

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 2]
```diff
@@ -328,64 +329,67 @@ public void testAllocateActionOnlyReplicas() throws Exception {
...
+        assertBusy(() -> assertThat(getFailedStepForIndex(index), equalTo("wait-for-snapshot")), slmPolicy);
 
         String repo = createSnapshotRepo();
-        createSlmPolicy(smlPolicy, repo);
+        createSlmPolicy(slmPolicy, repo);
 
-        assertBusy(() -> assertThat(getStepKeyForIndex(index).getAction(), equalTo("wait_for_snapshot")));
+        assertBusy(() -> assertThat(getStepKeyForIndex(index).getAction(), equalTo("wait_for_snapshot")), slmPolicy);
+        //wait for step to notice that the slm policy is created and to get out of error
+        assertBusy(() -> assertNull(getFailedStepForIndex(index)), slmPolicy);
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: Same as below, these could be grouped together as the are assertions of the same index state (ie. a particular action is the current action and the failed step info is missing)

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 3]
```diff
@@ -328,64 +329,67 @@ public void testAllocateActionOnlyReplicas() throws Exception {
...
             .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0));
-        String smlPolicy = randomAlphaOfLengthBetween(4, 10);
-        createNewSingletonPolicy("delete", new WaitForSnapshotAction(smlPolicy));
+        String slmPolicy = randomAlphaOfLengthBetween(4, 10);
+        createNewSingletonPolicy("delete", new WaitForSnapshotAction(slmPolicy));
         updatePolicy(index, policy);
-        assertBusy(() -> assertThat(getStepKeyForIndex(index).getAction(), equalTo("wait_for_snapshot")));
-        assertBusy(() -> assertThat(getFailedStepForIndex(index), equalTo("wait-for-snapshot")));
+        assertBusy(() -> assertThat(getStepKeyForIndex(index).getAction(), equalTo("wait_for_snapshot")), slmPolicy);
+        assertBusy(() -> assertThat(getFailedStepForIndex(index), equalTo("wait-for-snapshot")), slmPolicy);
```
Comment (andreidan) [inline x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: Same as below

**Comment by probakowski** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 1 above)
Comment (probakowski) [reply x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: Good idea, thanks!

**Pull request merged**